encrypted-media-respec.html

```diff
@@ -605,8 +605,112 @@
+            <p>Given a <a def-id="keysystems"></a> implementation <var>implementation</var>, <a>MediaKeySystemConfiguration</a> <var>candidate configuration</var>, and <var>origin</var>, this algorithm returns a supported configuration or <code>NotSupported</code> as appropriate.</p>
+            <p class="note">Unrecognized dictionary members in <var>candidate configuration</var> are ignored per [[WebIDL]] and will never reach this algorithm. Thus, they cannot be considered as part of the configuration.
+            </p>
+            <div class="note">
+              <p>
+              For certain configurations, it may be required to obtain user consent or inform the user. User Agents have some flexibility to determine
+              whether consent is required for a specific configuration and whether such consent may also apply to other configurations. For example,
+              consent to one configuration may also imply consent for less powerful, more restricted configurations. Equally, a denial of consent for
+              one configuration may imply denial of consent for more powerful, less restricted configurations.
+              </p>
+              <p>
+              Supported configurations, including supported audio and video codecs, may depend on availability of optional capabilities such as
+              <a def-id="distinctive-identifier"></a>s and persistent state. The following algorithm iteratively tries configurations with different
+              requirement values for these optional configuration elements until consent is achieved.
+              </p>
+              <p>
+              User Agents should persist consent responses at least for the duration of the <a def-id="requestMediaKeySystemAccess"></a>
+              algorithm in order to avoid repeated requests to the user for similar configurations.
+              </p>
+            </div>
+            <ol>
+              <li>
+                <p>Let <var>requirements</var> be the list [ <code>distinctiveIdentifier</code>, <code>persistentState</code> ].</p>
+              </li>
+              <li>
+                <p>
+                  Let <var>optional requirements</var> be the sublist of <var>requirements</var> where for each element of the sublist,
+                  the value of the corresponding member of <var>candidate configuration</var> is <a def-id="requirement-optional"></a>.
+                </p>
+              </li>
+              <li>
+                <dl class="switch">
+                  <dt>If <var>optional requirements</var> is not empty:</dt>
+                  <dd>
+                    <ol>
+                      <li>
+                        <p>
+                          Let <var>optional requirement combinations</var> be the list of all sublists of <var>optional requirements</var>
```

Comment:
Yes, this is what I meant by the generic version perhaps being overkill. Since we only have two MediaKeyRequirement there are only 4 possibilities, so we could list them explicitly. If we add another one we need the generic algorithm.

The ordering is important. We are ordering a list of lists, where the elements in each list are the ones which will be left as "optional". So we start leaving them all as optional (the longest list = the strongest requirement). If that fails, we try with exactly one of the optional items marked down to not-allowed. The order in which we try marking them down is the order we give for them at the start (in _requirements_).

At the moment we will try { distintiveIdentifier: not-allowed, persistentStore: optional } before we try { distintiveIdentifier: optional, persistentStore: not-allowed }. We could do it the other way around.

I'm not sure there is a definitive way to determine what is the best order: the important thing will be to have the order consistent across UAs. If the site wants to test these in a different order it can list them explicitly, without using the "optional" value.

Anyway, I'm happy to use a simpler algorithm. Or I could spell out the steps here in more detail and provide an example.